qutip/tests/core/test_qobjevo.py

```diff
@@ -42,111 +42,442 @@
+                _assert_qobj_almost_eq(copy(t), expected)
+            elif args_form in self.mixed_forms:
+                expected = (self.cte_qobj +
+                            self.cplx_coeff(t, args) * self.cplx_qobj +
+                            self.real_coeff(t, args) * self.real_qobj)
+                _assert_qobj_almost_eq(copy(t), expected)
+
+    @pytest.mark.parametrize('bin_op', [_add, _sub, _mul, _matmul, _tensor],
+                             ids=['add', 'sub', 'mul', 'matmul', 'tensor'])
+    def test_binopt(self, form, extra_form, bin_op):
+        "QobjEvo arithmetic"
+        t = self._rand_t()
+        obj1 = self.qobjevos[form]
+        obj1_t = self.qobjevos[form](t)
+        obj2 = self.qobjevos[extra_form]
+        obj2_t = self.qobjevos[extra_form](t)
+
+        as_qevo = bin_op(obj1, obj2)(t)
+        as_qobj = bin_op(obj1_t, obj2_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    @pytest.mark.parametrize('bin_op', [_add, _sub, _mul, _matmul, _tensor],
+                             ids=['add', 'sub', 'mul', 'matmul', 'tensor'])
+    def test_binopt_qobj(self, form, bin_op):
+        "QobjEvo arithmetic"
+        t = self._rand_t()
+        obj = self.qobjevos[form]
+        obj_t = self.qobjevos[form](t)
+        qobj = rand_herm(self.N)
+
+        as_qevo = bin_op(obj, qobj)(t)
+        as_qobj = bin_op(obj_t, qobj)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+        as_qevo = bin_op(qobj, obj)(t)
+        as_qobj = bin_op(qobj, obj_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    @pytest.mark.parametrize('bin_op', [_add, _sub, _mul, _div],
+                             ids=['add', 'sub', 'mul', 'div'])
+    def test_binopt_scalar(self, form, bin_op):
+        "QobjEvo arithmetic"
+        t = self._rand_t()
+        obj = self.qobjevos[form]
+        obj_t = self.qobjevos[form](t)
+        scalar = np.random.rand() + 1j
+
+        as_qevo = bin_op(obj, scalar)(t)
+        as_qobj = bin_op(obj_t, scalar)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+        if bin_op is not _div:
+            as_qevo = bin_op(scalar, obj)(t)
+            as_qobj = bin_op(scalar, obj_t)
+            _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    @pytest.mark.parametrize('unary_op', [_neg, _dag, _conj, _trans, _cdc],
+                             ids=['neg', 'dag', 'conj', 'trans', '_cdc'])
+    def test_unary(self, form, extra_form, unary_op):
+        "QobjEvo arithmetic"
+        t = self._rand_t()
+        obj = self.qobjevos[form]
+        obj_t = self.qobjevos[form](t)
+
+        as_qevo = unary_op(obj)(t)
+        as_qobj = unary_op(obj_t)
+        _assert_qobj_almost_eq(as_qevo, as_qobj)
+
+    def test_tidyup(self):
+        "QobjEvo tidyup"
+        obj = self.qobjevos["no_args"]
+        obj *= 1e-10 * np.random.random()
+        obj.tidyup(atol=1e-8)
+        t = self._rand_t()
+        # check that the Qobj are cleaned
+        assert_equal(np.max(obj(t).full()), 0.)
```

Comment:
use `np.testing.assert_allclose(obj(t).full(), 0)` so we get proper reporting on a failure (`assert_equal` should never ever be used any more - in `pytest`, the bare `assert x == y` is much clearer)